Fix setup upgrade command re enables caches

AbdulRahmanAbouzaid opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Fix re-enables all frontend caches after running setup:upgrade command including the disabled one BY Return them back to their old state before running setup:upgrade.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#28186

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1.Disable one of/all frontend caches [ layout, block_html and full_page] (for example: run `bin/magento cache:disable layout`).
2. run `bin/magento setup:upgrade`.
3. Run `bin/magento cache:status` and you can find that layout cache disabled.

<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

AbdulRahmanAbouzaid committed:
re-disable caches that were disabled before setup:upgrade command

AbdulRahmanAbouzaid committed:
lowercase the un needed uppercase words

AbdulRahmanAbouzaid commented:
@magento run all tests

AbdulRahmanAbouzaid committed:
confirm changes for only upgrade command

AbdulRahmanAbouzaid commented:
@magento run Unit Tests, WebAPI Tests, Sample Data Tests EE, Sample Data Tests B2B, Integration Tests, Functional Tests EE, Functional Tests CE, Functional Tests B2B

AbdulRahmanAbouzaid committed:
Merge branch '2.4-develop' into fix_setup-upgrade_command_re-enables_caches

AbdulRahmanAbouzaid commented:
@magento run all tests

AbdulRahmanAbouzaid committed:
update code style

AbdulRahmanAbouzaid committed:
Merge branch 'fix_setup-upgrade_command_re-enables_caches' of https://github.com/AbdulRahmanAbouzaid/magento2 into fix_setup-upgrade_command_re-enables_caches

AbdulRahmanAbouzaid commented:
@magento run Static Tests

AbdulRahmanAbouzaid commented:
@magento run Integration Tests

AbdulRahmanAbouzaid commented:
@magento run Functional Tests B2B

AbdulRahmanAbouzaid commented:
@magento run all tests

VladimirZaets committed:
Merge branch '2.4-develop' into fix_setup-upgrade_command_re-enables_caches

VladimirZaets commented:
@magento run all tests

magento-engcom-team commented:
Hi @VladimirZaets, thank you for the review.
[ENGCOM-8029](https://jira.corp.magento.com/browse/ENGCOM-8029) has been created to process this Pull Request
:eight_spoked_asterisk: @VladimirZaets, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Alfa commented:
@VladimirZaets Could you put an appropriate label for test coverage?

ihor-sviziev commented:
@engcom-Alfa I believe it requires automatic test, but not sure how to test CLI commands correctly

ihor-sviziev commented on setup/src/Magento/Setup/Model/Installer.php:
```diff
@@ -947,6 +959,10 @@ public function installDataFixtures(array $request = [])
         $this->handleDBSchemaData($setup, 'data', $request);
         $this->log->log('Enabling caches:');
         $this->updateCaches(true, $frontendCaches);
+        if ($keepCacheStatuses && !empty($disabledCaches)) {
+            $this->log->log('Disabling pre-disabled caches:');
+            $this->updateCaches(false, array_keys($disabledCaches));
+        }
```
It seems to me really bad to enable caches and disable them again. I think it's better to calculate what caches need to be re-enabled, and enable only needed

ihor-sviziev replied:
Just fixed it myself

ihor-sviziev commented:
@AbdulRahmanAbouzaid could you also cover your changes with some kind of test?

ihor-sviziev committed:
Fix setup upgrade command re enables caches

- Do not touch disabled cache types
- Re-enable cache in case if exception happened in handleDBSchemaData

ihor-sviziev committed:
Merge branch '2.4-develop' into fix_setup-upgrade_command_re-enables_caches

ihor-sviziev commented:
@magento run all tests

ihor-sviziev committed:
Fix setup upgrade command re enables caches

Fix static test failure

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

AbdulRahmanAbouzaid commented:
@ihor-sviziev thanks for your updates. Is there any other changes should be made?

ihor-sviziev reviewed (changes requested):
Hi @AbdulRahmanAbouzaid,
Could you cover your changes with any type of automated tests in order to prevent appearing this issue in future?

AbdulRahmanAbouzaid commented:
@ihor-sviziev I'll check this

engcom-Charlie commented:
Hi @AbdulRahmanAbouzaid, do you have any updates about test coverage?
Thank you.

gabrieldagama commented:
The risk was set to `medium` due to: the suggested changes may affect a critical area.

engcom-Charlie commented:
@AbdulRahmanAbouzaid I'll try to continue with your PR.

AbdulRahmanAbouzaid commented:
@engcom-Charlie Thank you. 
Sorry for the late reply. I tried to cover this with test but i couldn't

engcom-Charlie committed:
Merge branch '2.4-develop' into fix_setup-upgrade_command_re-enables_caches

engcom-Charlie committed:
test coverage

engcom-Charlie commented:
@magento run all tests

ihor-sviziev reviewed (approved):
✔ Approved.

Failing tests looks not related to changes form this PR.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8029](https://jira.corp.magento.com/browse/ENGCOM-8029) has been created to process this Pull Request

ihor-sviziev committed:
Merge branch '2.4-develop' into fix_setup-upgrade_command_re-enables_caches

ihor-sviziev commented:
@magento run all tests

engcom-Bravo commented:
:heavy_check_mark: QA Passed

The cache status has not been changed after performing Upgrade

### Manual testing scenario
1. Disable _block_html_ _full_page_ _layout_ using the following command
```bin/magento cache:disable block_html full_page layout```
2. Run ```bin/magento cache:status``` and assert that following are set to _0_
```layout: 0```
```block_html: 0```
```full_page: 0 ```
3. Upgrade magento instance using ```bin/magento setup:upgrade```
4. Check the cache status using  ```bin/magento cache:status```

### Actual Result
The previously disabled caches are still in _Disabled_ status (0)
```
Current status:
                        config: 1
                        layout: 0
                    block_html: 0
                   collections: 1
                    reflection: 1
                        db_ddl: 1
               compiled_config: 1
                           eav: 1
         customer_notification: 1
            config_integration: 1
        config_integration_api: 1
                     full_page: 0
             config_webservice: 1
                     translate: 1
```
### Before switching to PR 
```
Current status:
                        config: 1
                        layout: 1
                    block_html: 1
                   collections: 1
                    reflection: 1
                        db_ddl: 1
               compiled_config: 1
                           eav: 1
         customer_notification: 1
            config_integration: 1
        config_integration_api: 1
                     full_page: 1
             config_webservice: 1
                     translate: 1
```
